src/main/java/org/xbill/DNS/lookup/LookupResult.java

```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: BSD-2-Clause
+package org.xbill.DNS.lookup;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+import org.xbill.DNS.Name;
+import org.xbill.DNS.Record;
+
+/** LookupResult instances holds the result of a successful lookup operation. */
+@ToString
+@EqualsAndHashCode
+public final class LookupResult {
+  /** An unmodifiable list of records that this instance wraps, may not be null but can be empty */
+  @Getter private final List<Record> records;
+  /** provides additional detail when records is empty */
+  @Getter private final AdditionalDetail additionalDetail;
+  /** tracks the different CNAME and DNAME aliases for this name */
+  @Getter private List<Name> aliases;
+
+  /**
+   * Construct an instance with the provided records.
+   *
+   * @param records a list of records to return, or null if there was no response
+   * @param additionalDetail additional detail on this response, such as the reason for records
+   *     being null.
+   */
+  public LookupResult(List<Record> records, AdditionalDetail additionalDetail, List<Name> aliases) {
+    this.records = Collections.unmodifiableList(new ArrayList<>(records));
+    this.aliases = aliases;
+    if (records.isEmpty() && additionalDetail == null) {
+      throw new IllegalArgumentException(
+          "Empty records can not be combined with null additionalDetail");
+    }
+    this.additionalDetail = additionalDetail;
```

Comment:
Guarantee that this is not null with the OK status (as Java cannot (yet) handle null in switch statements).